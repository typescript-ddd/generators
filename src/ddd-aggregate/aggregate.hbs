import { AggregateRoot, CreateEntityProps, UpdateEntityProps } from "@typescript-ddd/core";
import { {{pascal entityName}}Id } from "./{{lower entityName}}-id";
import { {{pascal entityName}}CreatedEvent, {{pascal entityName}}UpdatedEvent } from "../events";

/**
 * Represents the properties required to create {{noun entityName}}.
 */
export interface Create{{pascal entityName}}Props extends CreateEntityProps {
  {{#each properties}}
  {{#if isNullable}}
  {{name}}?: {{type}},
  {{else}}
  {{name}}: {{type}},
  {{/if}}
  {{/each}}
}

/**
 * Represents the properties required to update {{noun entityName}}.
 */
export interface Update{{pascal entityName}}Props extends UpdateEntityProps<{{pascal entityName}}Id> {
  {{#each properties}}
  {{#if isNullable}}
  {{name}}?: {{type}},
  {{else}}
  {{name}}: {{type}},
  {{/if}}
  {{/each}}
}

/**
 * Represents {{noun entityName}}.
 */
export class {{pascal entityName}} extends AggregateRoot<{{pascal entityName}}Id> {
  /**
   * Initilizes a new instance of {{noun entityName}}.
   * @param {{lbrace}}{{entityName}}Id{{rbrace}} id - The unique identifier of the {{natural entityName}}.
   {{#each properties}}
   * @param {{lbrace}}{{type}}{{rbrace}} {{name}} - The {{name}} of the {{natural ../entityName}}.
   {{/each}}
   */
  private constructor(
    id?: {{pascal entityName}}Id,    
    {{#each properties}}
    {{#if isNullable}}
    private readonly {{name}}?: {{type}},
    {{else}}
    private readonly {{name}}: {{type}},
    {{/if}}
    {{/each}}) {
      super(id ?? {{pascal entityName}}Id.create());
      // TODO : Add validation

      if (!id) {
        this.applyChange(new {{pascal entityName}}CreatedEvent(this.id, this));
      }
    }

  {{#each properties}}
  /**
   * Gets the {{name}}.
   */
   get{{pascal name}}(): {{type}}{{#if isNullable}} | undefined{{/if}} {
     return this.{{name}};
   }

  {{/each}}
  /**
   * Updates {{noun entityName}}.
   * @param {Update{{pascal entityName}}Props} props - The properties to update the {{natural entityName}}.
   * @returns {void}
   */
  update(props: Update{{pascal entityName}}Props): void {
    // TODO : Add validation here
    {{#each properties}}
    this.{{name}} = props.{{name}};
    {{/each}}
    this.applyChange(new {{pascal entityName}}UpdatedEvent(this.id, props));
  }

  /**
   * Creates a new instance of {{noun entityName}}.
   * @param {{lbrace}}Create{{pascal entityName}}Props{{rbrace}} props - The properties required to create {{noun entityName}}.
   * @param {{lbrace}}{{pascal entityName}}Id{{rbrace}} id - Optional unique identifier of the {{natural entityName}}.
   * @returns {{lbrace}}{{pascal entityName}}{{rbrace}} - The newly created {{natural entityName}}.
   */
  static create(props: Create{{pascal entityName}}Props, id?: {{pascal entityName}}Id): {{pascal entityName}} {
    return new {{pascal entityName}}(
      id,
      {{#each properties}}
      props.{{name}},
      {{/each}}
    );
  }
}