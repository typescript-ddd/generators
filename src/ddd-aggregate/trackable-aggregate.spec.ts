import { AggregateTemplateData } from "./aggregate.types";
import { generateAggregate } from "./aggregate.generator";

describe('TrackableAggregate', () => {


    it('should render trackable template', () => {
        // Arrange
        const data:AggregateTemplateData = {
        entityName: 'User',
        properties: [
            { name: 'name', type: 'string' },
        ],
        trackable: true
        };
    
        // Act
        const result = generateAggregate(data);
    
        // Assert
        expect(result).toMatchSnapshot();
    });
});