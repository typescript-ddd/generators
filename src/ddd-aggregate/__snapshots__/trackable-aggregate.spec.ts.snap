// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackableAggregate should render trackable template 1`] = `
"// User.ts
import { TrackableAggregateRoot, CreateEntityProps, UpdateEntityProps, UtcDate } from "@typescript-ddd/core";
import { UserId } from "./user-id";
import { UserCreatedEvent, UserUpdatedEvent } from "../events";

/**
 * Represents the properties required to create a user.
 */
export interface CreateUserProps extends CreateEntityProps {
  name: string,
  createdAtUtc?: UtcDate,
  updatedAtUtc?: UtcDate,
}

/**
 * Represents the properties required to update a user.
 */
export interface UpdateUserProps extends UpdateEntityProps<UserId> {
  name: string,
}

/**
 * Represents a user.
 */
export class User extends TrackableAggregateRoot<UserId> {
  /**
   * Initilizes a new instance of a user.
   * @param {UserId} id - The unique identifier of the user.
   * @param {string} name - The name of the user.
   * @param {UtcDate} createdAtUtc - The date and time the user was created.
   * @param {UtcDate} updatedAtUtc - The date and time the user was last updated.
   */
  private constructor(
    id?: UserId,    
    private readonly name: string,
    createdAtUtc?: UtcDate,
    updatedAtUtc?: UtcDate) {
      super(id ?? UserId.create(), createdAt, updatedAt);
      // TODO : Add validation
      
      if (!id) {
        this.applyChange(new UserCreatedEvent(this.id, this));
      }
    }

  /**
   * Gets the name.
   */
   getName(): string {
     return this.name;
   }

  /**
   * Updates a user.
   * @param {UpdateUserProps} props - The properties to update the user.
   * @returns {void}
   */
  update(props: UpdateUserProps): void {
    // TODO : Add validation here
    this.name = props.name;
    this.applyChange(new UserUpdatedEvent(this.id, props));
  }

  /**
   * Creates a new instance of a user.
   * @param {CreateUserProps} props - The properties required to create a user.
   * @param {UserId} id - Optional unique identifier of the user.
   * @returns {User} - The newly created user.
   */
  static create(props: CreateUserProps, id?: UserId): User {
    return new User(
      id,
      props.name,
      props.createdAtUtc,
      props.updatedAtUtc,
    );
  }
}"
`;
